### PR TITLE
AP_Vehicle: add virtual getters to access basic waypoint info

### PR DESCRIPTION
**Pull request by yaapu**
This adds 3 pure virtual functions to AP_Vehicle to access:
- distance to next waypoint
- current waypoint bearing
- current waypoint crosstrack error

the implementations use the OSD code to be consistent across vehicles.

~~it's a POC to see if this approach is accepted~~

~~This also removes the need for the publish_osd_info call from the vehicle tasks~~

### CODE REVIEW

File: APMrover2/Rover.cpp [hunk 1]
```diff
@@ -338,6 +338,54 @@ void Rover::publish_osd_info()
...
+{
+    distance = control_mode->get_distance_to_destination();
+    return true;
+}
+
+/*
+  get the current wp bearing
+  return false if failed or n/a
+ */
+bool Rover::get_wp_bearing_cd(int32_t &bearing)
```
Comment (peterbarker) [inline APMrover2/Rover.cpp]: float degrees would be better.

File: APMrover2/Rover.cpp [hunk 2]
```diff
@@ -338,6 +338,54 @@ void Rover::publish_osd_info()
 }
 #endif
 
+/*
+  get the current wp index
+  return false if failed or n/a
+ */
+bool Rover::get_wp_index(uint16_t &index)
+{
+    if (control_mode == &mode_auto) {
```
Comment (peterbarker) [inline APMrover2/Rover.cpp]: It's a little odd that Rover cares about the mode but Copter does not.

File: ArduPlane/ArduPlane.cpp [hunk 3]
```diff
@@ -699,4 +699,44 @@ void Plane::publish_osd_info()
 }
 #endif
 
+/*
+  get the current wp index
+  return false if failed or n/a
+ */
+bool Plane::get_wp_index(uint16_t &index)
```
Comment (peterbarker) [inline ArduPlane/ArduPlane.cpp]: We could have a base-class implementation for get_wp_index, at the very least.

File: ArduSub/ArduSub.cpp [hunk 4]
```diff
@@ -345,4 +345,40 @@ bool Sub::control_check_barometer()
     return true;
 }
 
+/*
+  get the current wp index
+  return false if failed or n/a
+ */
+bool Sub::get_wp_index(uint16_t &index)
```
Comment (peterbarker) [inline ArduSub/ArduSub.cpp]: We `send_nav_controller_output` for Sub - can't see why we shouldn't have implementations here.

Comment (yaapu) [reply ArduSub/ArduSub.cpp]: by playing with the code it seems that a simple

uint16_t get_wp_index() is way simpler to handle than bool get_wp_index(uint16_t &index).
All methods called by this function are already designed to either return 0 or the correct value, there's no handling of a true/false result

Comment (peterbarker) [reply ArduSub/ArduSub.cpp]: And if `AP_Mission` isn't compiled in?

0 is a valid index.

**Comment by rmackay9** [comment]
Looks pretty OK but I'd skip the get_wp_index() and just directly access the mission library maybe.  i'm looking at it from the point of view of scripting and there's a PR from PeterHall that adds scripting access to the mission library I think.

**Comment by yaapu** [comment]
@rmackay9 yes, I did try to use AP_Mission::get_current_nav_index() but got 0 for both Copter and Plane, I might have done something wrong so I'll give it one more try.
@peterbarker there's a dependency for each vehicle on the osd object for OSD parameters access, another good candidate to a move to AP_Vehicle

I update the OSD nav info @2Hz, used to be @1Hz when scheduled from vehicle tasks, any feedback on this?

**Comment by yaapu** [comment]
@peterbarker I need help on how to get nav controller info for the sub vehicle

**Comment by yaapu** [comment]
@rmackay9 AP_Mission::get_current_nav_index() working fine now

**Comment by peterbarker** [comment]
> @rmackay9 yes, I did try to use AP_Mission::get_current_nav_index() but got 0 for both Copter and Plane, I might have done something wrong so I'll give it one more try.

Sorry, I disagree with having this return the index (sorry @rmackay9 !).  Sometimes we fill in our target lat/lon without having a waypoint number.  This is true in the nav-controller-output mavlink message where we cheerfully send through our loiter target (the output from the wpnav library) even though mission is not involved.  In that case we should not be associating where we are going with a waypoint number from the mission library as it could be misleading.

> @peterbarker there's a dependency for each vehicle on the osd object for OSD parameters access, another good candidate for a move to AP_Vehicle

Absolutely.  @andyp1per 's RunCam stuff sets precedent for that.

> I update the OSD nav info @2hz, used to be @1Hz when scheduled from vehicle tasks, any feedback on this?

This could instead be scheduled from the AP_Vehicle scheduler table.  We should probably discuss this as a team, but my feeling is for the time being AP_Vehicle should continue to push stuff into the OSD library rather than the library getting the data from AP_Vehicle - at least on this PR.  So essentially `publish_osd_info` moves up.

**Comment by yaapu** [comment]
> Sorry, I disagree with having this return the index (sorry @rmackay9 !). Sometimes we fill in our target lat/lon without having a waypoint number. This is true in the nav-controller-output mavlink message where we cheerfully send through our loiter target (the output from the wpnav library) even though mission is not involved. In that case we should not be associating where we are going with a waypoint number from the mission library as it could be misleading.

@peterbarker  in the case you raise what do you expect the osd to show on screen wp_idx = 0 or wp_idx = n/a or? Aren't we talking about a central source for telemetry info to be shared by scripting, osd and generic telemetry libraries? Or the ultimate goal is to have THE main source for such info and have mavlink ask ap_vehicle for infos?
I'll create an ap_vehicle member function based, just for the moment, on ap_mission::get_current_nav_index(), so it will be easy to change it once the team has made a decision on how it should be

> This could instead be scheduled from the AP_Vehicle scheduler table. We should probably discuss this as a team, but my feeling is for the time being AP_Vehicle should continue to push stuff into the OSD library rather than the library getting the data from AP_Vehicle - at least on this PR. So essentially publish_osd_info moves up.

ok, I can move this to a task

**Comment by peterbarker** [comment]
Please don't consider my ramblings the end answer to all of this :-)

Essentially, if you want to know what the current waypoint number is for auto mode, ask the Mission library.  (If it has 0 waypoints uploaded... well....) .

If you want to know which waypoint the vehicle is currently using - that's what this method is for.

It *might* be that you do want to display auto mode's waypoint number even if you are in `STABILIZE`.  MAVProxy does that.  In that case, you'd call the library's method.

**Comment by yaapu** [comment]
> Essentially, if you want to know what the current waypoint number is for auto mode, ask the Mission library. (If it has 0 waypoints uploaded... well....) .
> If you want to know which waypoint the vehicle is currently using - that's what this method is for.

basically if I want to know the current waypoint for a non auto mode I should ask this method which has no way to know but to as each vehicle?

File: APMrover2/Rover.cpp (hunk 2 above)
Comment (yaapu) [reply APMrover2/Rover.cpp]: this was to be consistent with how Rover::publish_osd_info() worked vs Copter::publish_osd_info()

File: Rover/Rover.cpp [hunk 5]
```diff
@@ -384,23 +381,31 @@ void Rover::update_mission(void)
...
+bool Rover::get_wp_distance_m(float &distance) const
+{
+    distance = control_mode->get_distance_to_destination();
+    return true;
+}
+
+// vehicle specific waypoint info helpers
+bool Rover::get_wp_bearing_d(float &bearing) const
+{
+    bearing = control_mode->wp_bearing() * 100.0f;
```
Comment (tridge) [inline Rover/Rover.cpp:394]: should be 0.01 ?

File: ArduSub/ArduSub.cpp [hunk 6]
```diff
@@ -326,4 +326,26 @@ bool Sub::control_check_barometer()
     return true;
 }
 
+// vehicle specific waypoint info helpers
+bool Sub::get_wp_distance_m(float &distance) const
+{
+    distance = sub.wp_nav.get_wp_distance_to_destination() * 1.0e-2f;
```
Comment (tridge) [inline ArduSub/ArduSub.cpp:332]: prefer 0.01

File: ArduSub/ArduSub.cpp [hunk 7]
```diff
@@ -326,4 +326,26 @@ bool Sub::control_check_barometer()
...
+    return true;
+}
+
+// vehicle specific waypoint info helpers
+bool Sub::get_wp_bearing_d(float &bearing) const
+{
+    bearing = sub.wp_nav.get_wp_bearing_to_destination() * 1.0e-2f;
```
Comment (tridge) [inline ArduSub/ArduSub.cpp:339]: 0.01

File: ArduCopter/Copter.cpp [hunk 8]
```diff
@@ -617,17 +614,26 @@ void Copter::update_altitude()
...
+bool Copter::get_wp_bearing_d(float &bearing) const
+{
+    bearing = flightmode->wp_bearing();
+    return true;
+}
+
+// vehicle specific waypoint info helpers
+bool Copter::get_wp_crosstrack_error_m(float &xtrack_error) const
+{
+    xtrack_error = flightmode->crosstrack_error() * 1.0e-2f;
```
Comment (tridge) [inline ArduCopter/Copter.cpp:634]: 0.01

File: ArduCopter/Copter.cpp [hunk 9]
```diff
@@ -617,17 +614,26 @@ void Copter::update_altitude()
...
+// vehicle specific waypoint info helpers
+bool Copter::get_wp_distance_m(float &distance) const
 {
-    AP_OSD::NavInfo nav_info;
-    nav_info.wp_distance = flightmode->wp_distance() * 1.0e-2f;
-    nav_info.wp_bearing = flightmode->wp_bearing();
-    nav_info.wp_xtrack_error = flightmode->crosstrack_error() * 1.0e-2f;
-    nav_info.wp_number = mode_auto.mission.get_current_nav_index();
-    osd.set_nav_info(nav_info);
+    distance = flightmode->wp_distance() * 1.0e-2f;
```
Comment (tridge) [inline ArduCopter/Copter.cpp:620]: 0.01

File: ArduCopter/Copter.cpp [hunk 10]
```diff
@@ -617,17 +614,26 @@ void Copter::update_altitude()
...
+    return true;
+}
+
+// vehicle specific waypoint info helpers
+bool Copter::get_wp_bearing_d(float &bearing) const
+{
+    bearing = flightmode->wp_bearing();
```
Comment (tridge) [inline ArduCopter/Copter.cpp:627]: times 0.01 ?

File: Rover/Rover.h [hunk 11]
```diff
@@ -390,9 +390,10 @@ class Rover : public AP_Vehicle {
     bool should_log(uint32_t mask);
     bool is_boat() const;
 
-#if OSD_ENABLED || OSD_PARAM_ENABLED
-    void publish_osd_info();
-#endif
+    // vehicle specific waypoint info helpers
+    bool get_wp_distance_m(float &distance) const override;
+    bool get_wp_bearing_d(float &bearing) const override;
```
Comment (tridge) [inline Rover/Rover.h:395]: rename to _deg()

File: libraries/AP_Vehicle/AP_Vehicle.h [hunk 12]
```diff
@@ -214,6 +214,24 @@ class AP_Vehicle : public AP_HAL::HAL::Callbacks {
...
+      get the distance to next wp in meters
+      return false if failed or n/a
+     */
+    virtual bool get_wp_distance_m(float &distance) const { return false; }
+
+    /*
+      get the current wp bearing in degrees
+      return false if failed or n/a
+     */
+    virtual bool get_wp_bearing_d(float &bearing) const { return false; }
```
Comment (tridge) [inline libraries/AP_Vehicle/AP_Vehicle.h:227]: _deg()

**Review by tridge — CHANGES_REQUESTED** [review]

File: Rover/Rover.cpp (hunk 5 above)
Comment (yaapu) [reply Rover/Rover.cpp]: control_mode->wp_bearing() it's already in degrees so no multiplier (OSD expects bearing in centidegrees though)

File: ArduSub/ArduSub.cpp (hunk 6 above)
Comment (yaapu) [reply ArduSub/ArduSub.cpp]: fixed

File: ArduSub/ArduSub.cpp (hunk 7 above)
Comment (yaapu) [reply ArduSub/ArduSub.cpp]: fixed

File: ArduCopter/Copter.cpp (hunk 8 above)
Comment (yaapu) [reply ArduCopter/Copter.cpp]: fixed

File: ArduCopter/Copter.cpp (hunk 9 above)
Comment (yaapu) [reply ArduCopter/Copter.cpp]: fixed

File: ArduCopter/Copter.cpp (hunk 10 above)
Comment (yaapu) [reply ArduCopter/Copter.cpp]: fixed

File: Rover/Rover.h (hunk 11 above)
Comment (yaapu) [reply Rover/Rover.h]: fixed

File: libraries/AP_Vehicle/AP_Vehicle.h (hunk 12 above)
Comment (yaapu) [reply libraries/AP_Vehicle/AP_Vehicle.h]: fixed

**Comment by yaapu** [comment]
@peterbarker @tridge I checked and tested all units, virtual getters return degrees or meters, OSD code expects bearing in cd so a conversion is needed.
I made all getters consistent with their respective GCS_MAVLINK_<*>::send_nav_controller_output(), so for instance Plane now checks if quadplane is in vtol mode.

File: ArduPlane/ArduPlane.cpp [hunk 13]
```diff
@@ -601,17 +598,49 @@ float Plane::tecs_hgt_afe(void)
...
+}
+
+bool Plane::get_wp_crosstrack_error_m(float &xtrack_error) const
+{
+    // see GCS_MAVLINK_Plane::send_nav_controller_output()
+    if (control_mode == &mode_manual) {
+        return false;
+    }
+    if (quadplane.in_vtol_mode()) {
+        xtrack_error = quadplane.using_wp_nav() ? quadplane.wp_nav->crosstrack_error() : 0;
```
Comment (peterbarker) [inline ArduPlane/ArduPlane.cpp:637]: Unrelated to your PR, but `using_wp_nav` could possibly do with a `quadplane.wp_nav != nullptr` in it.

**Review by peterbarker — APPROVED** [review]
LGTM

**Comment by peterbarker** [comment]
Merged, thanks @yaapu !

**Pull request merged**